uni-stark/src/folder.rs

```diff
@@ -63,8 +65,9 @@ impl<'a, SC: StarkGenericConfig> AirBuilder for ProverConstraintFolder<'a, SC> {
     #[inline]
     fn assert_zero<I: Into<Self::Expr>>(&mut self, x: I) {
         let x: PackedVal<SC> = x.into();
-        self.accumulator *= self.alpha_repeated;
-        self.accumulator += x;
+        let alpha_power = self.alpha_powers[self.constraint_index];
+        self.accumulator += PackedChallenge::<SC>::from_f(alpha_power) * x;
+        self.constraint_index += 1;
```

Comment:
It's out of scope for this PR but, if this is still slow, there is a possible (almost) 2x speed up possible here for MONTY fields.

Unfortunately it would require a little re-engineering as it would involve adding everything in pairs. It makes use of the fact that there is a really efficiently algorithm for a*b + c*d for these fields.

Reply:
Oh interesting, is it a form of amortizing the Monty reduction? Could we amortize it over even more terms, maybe 4 terms so the dot product fits in 64 bits?

Not sure if the APIs will complicate things (since we're dealing with an extension over a packed field) but it sounds great in principle.

Reply:
Yeah essentially. The point is that MONTY reduction works provided the input is <= 2^31 P ~ 2P^2 so to compute `ab + cd` we can compute this as a 64 bit element and do a single Monty reduction.

So we end up needing 2 integer mul's, 1 integer add and 1 monty reduction compared to 2 integer mul's, 2 monty reductions and 1 field add. As the cost of multiplication is dominated by monty reductions this is basically twice as fast.

If we wanted to do more terms this would be a little more complicated but should be doable. It would probably end up being a combination of the idea here and Jacqui's fast dot product code see: #252 

The main pain point in order to do this is that we probably need to modify API so that this is actually a dot product as opposed to the current method.

Reply:
Paper: https://eprint.iacr.org/2022/367

Implementation: https://github.com/mratsim/constantine/blob/2c3de87/constantine/math/arithmetic/limbs_montgomery.nim#L308-L363. Important I assume that there are 2 spare bits (so 2^30 at most or 2^62 at most for small fields): for:
- delayed carry according to https://hackmd.io/@gnark/modular_multiplication (this only matters for big fields)
- skipping final substraction according to https://eprint.iacr.org/2017/1057.pdf

Reply:
Hmm, I might be misunderstanding but don't these methods mainly apply to cases where the prime modulus p is very large? (e.g. `p ~ 2^128, 2^256, 2^512`) Whereas in our case where `p ~ 2^31` fits in a single scalar register/packs nicely into vector registers?

Might be worth creating an issue or adding comments under #252.

Reply:
They are applicable as long as you have some spare bits, say you use 31 bits out of the physical register 32 bits.